flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -530,15 +530,22 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
 							request.getOnCompletionFuture()),
 						timer);
 
-			final CompletableFuture<?> masterStatesComplete = pendingCheckpointCompletableFuture
-					.thenCompose(this::snapshotMasterState);
-
 			final CompletableFuture<?> coordinatorCheckpointsComplete = pendingCheckpointCompletableFuture
 					.thenComposeAsync((pendingCheckpoint) ->
 							OperatorCoordinatorCheckpoints.triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(
 									coordinatorsToCheckpoint, pendingCheckpoint, timer),
 							timer);
 
+			// We have to take the snapshot of the master hooks after the coordinator checkpoints has completed.
+			// This is to ensure the tasks are checkpointed after the OperatorCoordinators in case
+			// ExternallyInducedSource is used.
+			final CompletableFuture<?> masterStatesComplete = coordinatorCheckpointsComplete
+					.thenComposeAsync(ignored -> {
+						PendingCheckpoint checkpoint =
+							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
```

Comment:
Ok, I see how it's supposed to be working. But in that case there is a missing `checkState` here.

```
PendingCheckpoint checkpoint =
	FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);

Preconditions.checkState(
	checkpoint != null || throwable != null,
	"?????????.");
```
% maybe code deduplication with the same get and check from  L553:L558. It's better to fail early, with `IllegalStateException` compared to `NPE` in `snapshotMasterState`.

Reply:
> the behavior is guaranteed by CompletableFuture, the assertion here would essentially be verifying CompletableFuture,

Not exactly. It would be verifying that you chained a couple of futures and callbacks correctly. That the callback `foo()` is using `future1` result and is triggered once `future2` completes, and that `future1` and `future2` are chained (or am I still mis understanding this code?). Java library doesn't guarantee you that, but your code that is chaining the futures does. Which is outside of the `foo()`'s control, so from `foo()`s perspective, that's an external assumption, and falls under:

> ensure the interface contract with users are not broken.

Where "users" are function's callers.

And as I wrote before. If something violates this assumption, and even if some unit test fail, it's a bit easier to understand a `checkState` compared to `NPE`. Note performance overhead of one if check doesn't matter here at all. Also it's harder of `checkState` to become outdated and misleading over time.

If you have so strong feelings about, put a comment, but I do not see any drawback of replacing a comment with a `checkState` with the same comment but as a message here.